apps/emqx/src/emqx_persistent_session_fdb_backend.erl

```diff
@@ -0,0 +1,138 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2021 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_persistent_session_fdb_backend).
+
+-include("emqx.hrl").
+-include("logger.hrl").
+-include("emqx_persistent_session.hrl").
+
+-define(FDB_SESSION_STORE, emqx_fdb_session_store).
+
+-export([ create_tables/0
+        , first_message_id/0
+        , next_message_id/1
+        , delete_message/1
+        , first_session_message/0
+        , next_session_message/1
+        , delete_session_message/1
+        , put_session_store/1
+        , delete_session_store/1
+        , lookup_session_store/1
+        , put_session_message/1
+        , put_message/1
+        , get_message/1
+        , ro_transaction/1
+        ]).
+
+create_tables() ->
+    Db = erlfdb:open(),
+    persistent_term:put(?FDB_SESSION_STORE, {Db}).
+
+first_session_message() ->
+    {Db} = persistent_term:get(?FDB_SESSION_STORE),
+    Opts = [{limit, 1}],
+    case erlfdb:get_range_startswith(Db, erlfdb_tuple:pack({<<"SESS_MSG_TAB">>}), Opts) of
+        [] -> '$end_of_table';
+        [{_, Binary}] -> 
+            ?SLOG(debug, #{db => fdb, first_session_message => (binary_to_term(Binary))#session_msg.key}),
```

Comment:
Just do one `binary_to_term` and bind it to a variable. This can be costly.

Reply:
There are more places like this. I leave them uncommented.

Reply:
You're right.